### DISCUSSION TITLE
Setting theme variant

### DISCUSSION THREAD

**Discussion by hkparker** [fyne-io/fyne#5337] [Q&A] [answered]
I'd like to have a toggle where users can select between dark mode and light mode.  This is defined by the `fyne.ThemeVariant`.  I can control the value of the theme variant via an envar, but I haven't figured out how to do it programmatically.

I'm looking for something like `fyne.CurrentApp().Settings().SetThemeVariant(theme.VariantDark)`

It looks like I could define a custom theme, then in the calls to `Color` ignore Fyne's theme variant and have my own variable I use to track variant (ignoring Fyne's settings).  That would involve me having to define each color in there, as I wouldn't be able to pass things to the default theme as far as I can tell.  (right now I'd be happy keeping the default theme, I just want to be able to control the variant).

I feel like I must be missing something.  With the default theme, or with a custom theme, how should I be going about setting the variant on the fly?

**Comment by dweymouth** [comment, marked as answer]
You can delegate to the default theme from your custom theme for all the Color, Size, etc calls.

Take a look at how [fyne_demo](https://github.com/fyne-io/fyne/blob/4a875d9d5677ce8a98bd15787d6f74d4263cac16/cmd/fyne_demo/main.go#L244) does it, with a custom theme called `forcedVariant`, which just delegates to the builtin theme for everything except the variant value

**Reply by hkparker** [reply]
Oh nice, yeah this is super clean.  Kind of a silly question I see, this was easy to find, thank you for pointing me in the right direction.

It looks to me like this forced theme doesn't change the variant that is stored in settings.  So, I'm assuming I can use the setting's variant to check what the system wanted it to be, and use my forcedVariant theme for what the user wants (which is to say, if the system defaults to dark, and I set it to light, checking `Settings.ThemeVariant()` will still return dark).

**Reply by andydotxyz** [reply]
100% :)

**Reply by hkparker** [reply]
Ok cool.  A bit of an opportunity for confusion in that I think of settings as things that I can control, and the variant in there is more of a refection of the underlying system state, not a setting I can flip.  But that's not something I'd really push on changing (at least, not unless there's a struct designed to give you those sorts of system states, in which case maybe it would make more sense over there).

Anyway I'm off to the races on a custom theme now based on the forcedVariant theme, however I'm running into the issues that many of my custom widgets are not responsive to theme changes and, looking at the built in widgets for reference, I can't figure out why.  I don't have a specific question on that yet, I'm sure I'm missing something simple, maybe something I need to be doing in the renderers' Refresh calls.  But I'll probably take that to slack or another discussion if I remain stuck for a while.
